docs/rfcs/bun-bundler.ts

```diff
@@ -0,0 +1,235 @@
+
+export interface BundlerOptions {
+  // currently this only accepts a Loader string
+  // change: set an ext->loader map
+  loader?: { [k in string]: Loader };
+
+  // rename `platform` to `target`
+  target?:
+    | Target
+    | {
+        name: Target; // e.g. 'node', 'browser'
+        // instead of esbuild's `target`
+        version?: string; // e.g. '18', 'es2020'
+        // low priority
+        supported?: Record<string, boolean>;
+      };
+
+  // currently accepts a tsconfig as stringified JSON or an object
+  // IMO this should accept a file path to a tsconfig.json file instead
+  // or at least in addition
+  // that's what most people are used to in config files
+  tsconfig?: string;
+
+  // I don't know what this is
+  allowBunRuntime?: boolean;
+
+  ////////////////////////////////
+  // no changes below this line //
+  ////////////////////////////////
+  macro?: MacroMap;
+  autoImportJSX?: boolean;
+  trimUnusedImports?: boolean;
+  jsxOptimizationInline?: boolean;
+  inline?: boolean;
+
+  sourcemap?:
+    | boolean
+    | {
+        root?: string;
+        inline?: boolean;
+        external?: boolean;
+
+        // probably unnecessary
+        content?: boolean;
+      };
+
+  format?:
+    | Format
+    | {
+        type: "iife";
+        globalName?: string;
+      };
+
+  logging?: {
+    color?: boolean;
+    level?: LogLevel;
+    // extremely low priority
+    limit?: number;
+    override?: Record<string, LogLevel>;
+  };
+
+  // removed: mangleProps, reserveProps, mangleQuoted, mangleCache
+
+  /** Documentation: https://esbuild.github.io/api/#minify */
+  minify?:
+    | boolean
+    | {
+        whitespace?: boolean;
+        identifiers?: boolean;
+        syntax?: boolean;
+      };
+
+  treeShaking?:
+    | boolean
+    | {
+        imports?: boolean;
+        code?: boolean;
+        // whether to believe @__PURE__
+        // I hate this
+        pure?: boolean;
+        // whether to believe package.json sideEffects
+        sideEffects?: boolean;
+      };
+
+  jsx?:
+    | JsxTransform
+    | {
+        transform?: JsxTransform;
+        factory?: string;
+        fragment?: string;
+        importSource?: string;
+        development?: boolean;
+        sideEffects?: boolean;
+      };
+
+  /**
+   * Low priority
+   * */
+  legalComments?: "none" | "inline" | "eof" | "linked" | "external";
+  charset?: Charset;
+}
+
+// `.build(options: BuildOptions)`
+// can override everything specified in Bundler options
+interface BuildOptions extends BundlerOptions {
+  bundle?: boolean; // default true
+  splitting?: boolean;
+  plugins?: BunPlugin[];
+  cwd?: string;
+  watch?: boolean;
+
+  // dropped: preserveSymlinks
+  // defer to tsconfig for this
+
+  // whether to parse manifest after build
+  manifest?: boolean;
+
+  // inputs
+  entrypoints: string[];
+  rootDir?: string; // equiv. outbase
+
+  memory?: boolean;
+
+  // no outfile; it can be specified with `outdir` and `outnames`
+  write?:
```

Comment:
instead of this and `memory`

what if we synchronously return a `Build` object which is a lazily-evaluated then-able? 

```ts
interface Build {
   readonly outputFiles: Promise<OutputFile>,
   then(): Promise<Build>,
    
   write(dirOrFile: string | FileReference): Promise<void>;
   text(): Promise<string>;
   blob(): Promise<Blob>;
   arrayBuffer(): Promise<ArrayBuffer>;
}
```

also curious to hear what @privatenumber thinks

Reply:
Then each `OutputFile` would extend `Blob` and possibly have a `sourceMap`?

Reply:
That reminds me of Rollup's API, where they [separate bundling from asset generation](https://rollupjs.org/javascript-api/#rollup-rollup). There's performance benefits to this when you output multiple formats because it can re-use the bundling work, and I think this API design implies that. Would it be possible to persist & re-use the bundle data with Bun?

With esbuild, I think it doesn't need to because it's so fast (but I think it could still benefit in distributions that are slower e.g. WASM). And also, the binary is stateless and no longer long-running.

In either case, I agree memory/write can be merged to one property as they're mutually exclusive.

Reply:
> Would it be possible to persist & re-use the bundle data with Bun?

Probably not

> There's performance benefits to this

Yeah. In this case, the goal is to skip cloning the data if it's not strictly necessary. If they're going to send it over the network without reading the data as text or ArrayBuffer, we can skip cloning the bytes in the Response constructor. If they're immediately going to compress the data we can .gzip() it and possibly reuse memory while compressing

Reply:
Will Bun defer building until the methods are called? What is the first form of data it's stored as (I'm guessing `Blob` or `ArrayBuffer`/`Uint8Array` depending the Node.js version you want to support.)

I'm not sure if `texts()`, `blobs()`, `arrayBuffers()` make sense because they need file names, and they can be made redundant by defining `OutputFile` like this:
```ts
type OutputFile = {
	name: string;
	content: Blob; // or ArrayBuffer
}
```

where if the user wants the text, they can call [`Blob#text()`](https://developer.mozilla.org/en-US/docs/Web/API/Blob/text):
```ts
const build = build({ ... });
const files = await build.files();

for (const file of files) {
	console.log(
		file.name,
		await file.content.text(),
	);
}
```
(Would there be benefit for `Build#files()` to be an async generator?)

RE: `gzip()` In my experience, gzipping is usually a server concern so it's handled by some middleware. I don't often see gzip encoded files written to disk, but if that's necessary, I think they should import a gzip util.